---
title: "Using celda with RNASeq: TBD Analysis"
author: "Josh D. Campbell, Sean Corbett, Masanao Yajima, Zichun Liu, Shiyi Yang, Tianwen Huan, Anahita Bahri"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Overview

We will learn how to use the package `celda` to cluster single cell RNAseq experiments.

Starting from a matrix of counts of transcripts (cells by transcripts). we will discuss the preliminary steps of ________, __________, and ________. Once we are satisfied that we have high-quality expression, we will consider tests for __________ and ways to visualize results.  

We will apply these methods to a data set of __________, measured on _________. 

```{r,eval=T, warning = FALSE}
library(devtools)
library(celda)
library(Matrix)
library(gtools)
library(ggplot2)
```


## Loading and transforming data

We use the built-in data generating function `simulateCells.celda_CG` to simulate a single-cell RNA-Seq dataset. `simulateCells.celda_CG` creates a list containing the original cell cluster

```{r}
sim_counts = simulateCells.celda_CG()
str(sim_counts)
```


```{r fig.width = 7, fig.height = 7}
plot(sim_counts$counts, main = "Simulated Cell Counts", xlab = "gene", ylab = "cells")

```

## Normalize counts data

```{r fig.width = 7, fig.height = 7}
norm.counts <- normalizeCounts(sim_counts$counts)
plot(norm.counts, main = "Normalized Simulated Cell Counts", xlab = "gene", ylab = "cells")
```

## Exploratory Data Analysis

Letâ€™s explore these data with a heatmap and some PCA.

```{r, fig.show='hold', warning = FALSE, fig.width = 7, fig.height = 7}
celdaCG.res = celda(counts = norm.counts, model = "celda_CG", nchains = 2, K =10, L =10)

z = celdaCG.res$res.list[[2]]$z
y = celdaCG.res$res.list[[2]]$y
render_celda_heatmap(counts=norm.counts,z=z,y=y)
```
```{r, fig.show='hold', fig.width = 7, fig.height = 7}
z = t(scale(t(norm.counts)))
pc = prcomp(z, center = F, scale = F)
plot(pc$rotation[,1], pc$rotation[,2], 
    pch=21, col="black", bg="blue",
    xlab="PC1", ylab="PC2")
```


