---
title: "Using celda with RNASeq: TBD Analysis"
author: "Josh D. Campbell, Sean Corbett, Masanao Yajima, Zichun Liu, Shiyi Yang, Tianwen Huan, Anahita Bahri"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Overview

We will learn how to use the package `celda` to cluster single cell RNAseq experiments.

Starting from a matrix of counts of transcripts (cells by transcripts). we will discuss the preliminary steps of ________, __________, and ________. Once we are satisfied that we have high-quality expression, we will consider tests for __________ and ways to visualize results.  

We will apply these methods to a data set of __________, measured on _________. 

```{r,eval=T}
library(devtools)
library(celda)
library(Matrix)
library(gtools)
library(ggplot2)
```


We use the built-in data generating function `cCG.generateCells`


## Loading and transforming data

```{r}
sim_counts = cCG.generateCells()
str(sim_counts)
```


Sean/Josh: Can you add some description on `cCG.generateCells`?

Some illustrations.

```{r}
plot(1:10)
qplot(1:10,1:10)



```


## Exploratory Data Analysis

Letâ€™s explore these data with a heatmap and some PCA.

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```
```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```


## TBD

